### PR TITLE
Run AST comparison tests on Travis

### PR DESCRIPTION
**Pull request by josephfrazier**
It looks like some of these currently fail, so we should probably also
sort that out.

Inspired by https://github.com/prettier/prettier/issues/1552

### CODE REVIEW

**Comment by shaunlebron** [comment]
looks like `debug-check` (for cli) and `AST_COMPARE` (for tests) are doing different things and maybe need to be consolidated

**Comment by josephfrazier** [comment]
Yeah, I'm actually looking into that at the moment. I'm doing to see if making AST_COMPARE act like --debug-check fixes things. If so, I'll push that up as well.

**Comment by josephfrazier** [comment]
Alright, after making `AST_COMPARE=1` work like `--debug-check`, then using the correct parser for the AST comparisons, and doing a little more AST massaging to remove leading/trailing comments and the `extra` property, I've gotten the test failures down to actually relevant AST differences. You can see a summary of the failures (as of commit 6923a35) here: https://travis-ci.org/prettier/prettier/jobs/230151061#L2460

As you might expect, the failures can be replicated by running `prettier --debug-check --parser PARSER` on the relevant file. For example:

```sh
prettier --debug-check --parser flow tests/flow/getters_and_setters_enabled/object_type.js
```

**Comment by vjeux** [comment]
Awesome, thanks a lot for doing that work. Do you want to commit this without running it inside of travis so that it's in master and people can in parallel fix all the tests that are broken?

**Comment by josephfrazier** [comment]
Thanks for the quick response! Getting the non-Travis changes in first sounds great, I just pushed up a commit for that. Let me know if/how you'd like me to rebase before merging.

File: src/parser.js [hunk 1]
```diff
@@ -1,12 +1,41 @@
...
+    parseFunction = parseWithBabylon;
+  }
+
+  try {
+    return parseFunction(text);
+  } catch (error) {
+    const loc = error.loc;
+
+    if (loc) {
+      error.codeFrame = codeFrame(text, loc.line, loc.column + 1, {
```
Comment (vjeux) [inline src/parser.js]:
```js
var codeFrame = require("babel-code-frame");
```

File: src/parser.js [hunk 2]
```diff
@@ -1,12 +1,41 @@
 "use strict";
 
+const codeFrame = require("babel-code-frame");
```
Comment (vjeux) [inline src/parser.js]: since you are moving this, could you do an inline require, only require it when it's actually used.

File: index.js [hunk 3]
```diff
@@ -115,8 +87,10 @@ module.exports = {
   version: version,
   __debug: {
     parse: function(text, opts) {
-      return parse(text, opts);
+      return parser.parse(text, opts);
     },
+    cleanAST: parser.cleanAST,
+    massageAST: parser.massageAST,
```
Comment (vjeux) [inline index.js]: We don't want to expose those two functions in the public npm export. Not only they have a really bad name, but also people shouldn't really do those kind of things.

File: src/parser.js [hunk 4]
```diff
@@ -97,4 +126,76 @@ function isProbablyJsx(text) {
...
+    ) {
+      delete newObj.key;
+    }
+
+    return newObj;
+  }
+  return ast;
+}
+
+module.exports = { parse, parseWithFlow, parseWithBabylon, parseWithTypeScript, cleanAST, massageAST };
```
Comment (vjeux) [inline src/parser.js]: why do you export `parseWithBabylon` now that you moved `parse` inside. I don't think that you are using it anymore?

File: src/parser.js [hunk 5]
```diff
@@ -97,4 +126,76 @@ function isProbablyJsx(text) {
...
+      ast.key &&
+      (ast.key.type === "Literal" || ast.key.type === "Identifier")
+    ) {
+      delete newObj.key;
+    }
+
+    return newObj;
+  }
+  return ast;
+}
```
Comment (vjeux) [inline src/parser.js]: They don't have any dependency on parser.js, could you just create another file instead?

File: src/parser.js (hunk 4 above)
Comment (josephfrazier) [inline src/parser.js]: Whoops, I didn't think to check if it was used elsewhere. Fixed in https://github.com/prettier/prettier/pull/1553/commits/46c14b94cbe98270a45a71bf5e48732bec2458a6

File: src/parser.js [hunk 6]
```diff
@@ -97,4 +125,4 @@ function isProbablyJsx(text) {
   ].join(""), "m").test(text);
 }
 
-module.exports = { parseWithFlow, parseWithBabylon, parseWithTypeScript };
+module.exports = { parse };
```
Comment (vjeux) [inline src/parser.js]: just export back a function :)

File: src/parser.js (hunk 5 above)
Comment (josephfrazier) [inline src/parser.js]: Sure, is `src/clean-ast.js` ok? See https://github.com/prettier/prettier/pull/1553/commits/338b4f93bcc40679e2861ad07037e52d9f5459f2

**Comment by vjeux** [comment]
Awesome, thanks. Once travis is green i'll merge this!

**Comment by vjeux** [comment]
Looks like you are using some syntax that isn't supported in node 4 :(

File: tests_config/run_spec.js [hunk 7]
```diff
@@ -56,20 +45,22 @@ function run_spec(dirname, options, additionalParsers) {
 
       if (RUN_AST_TESTS) {
         const source = read(dirname + "/" + filename);
-        const ast = removeEmptyStatements(parse(source));
-        let ppast;
+        const ast = parse(source, mergedOptions);
+        const astMassaged = massageAST(ast);
```
Comment (vjeux) [inline tests_config/run_spec.js]: where is massageAST being defined?

Comment (josephfrazier) [inline tests_config/run_spec.js]: Heh, I got ahead of myself :P It should be good now, along with the Node 4 test.

**Comment by vjeux** [comment]
Thanks!

**Comment by josephfrazier** [comment]
Thanks for the merge! I'll go ahead and create issues for the failed tests (other than https://github.com/prettier/prettier/issues/1555)

**Pull request merged**